### PR TITLE
feat (binding-netconf): add support for NetConf binding

### PR DESCRIPTION
**Pull request by lukesmolo**
Add NetConf client protocol binding.
It currently supports the followings NetConf methods:

- GET-CONFIG
- EDIT-CONFIG
- COMMIT
- (CUSTOM) RPC

Subscriptions are still not implemented.

Signed-off-by: lukesmolo <lukesmolo@gmail.com>

### CODE REVIEW

**Comment by sebastiankb** [comment]
Thank you so much for this PR! May I ask you to provide a README about this binding and an example script + TD. That would be awesome.

One more question: what kind of (secured) transport protocol is supported in this binding?

**Comment by lukesmolo** [comment]
> Thank you so much for this PR! May I ask you to provide a README about this binding and an example script + TD. That would be awesome.

 Sure, no problem. I'll add them as soon as possible.
> One more question: what kind of (secured) transport protocol is supported in this binding?

This binding uses [node-netconf](https://www.npmjs.com/package/netconf), that relies on ssh for sending RPC requests. So I would answer that this binding supports TCP for the SSH connection. The security of the connection is then granted by the user/password or user/privateKey combinations provided in the servient configuration.

**Comment by lukesmolo** [comment]
Ok, @sebastiankb you can find now an example and the README as well.

**Comment by danielpeintner** [comment]
Question: you also changed `ProtocolClient` by adding schema information. Is this actually needed?

Moreover, if we change it we should use `TD.DataSchema` consistently and not mix `TD.DataSchema` and `TD.BaseSchema`.

**Comment by lukesmolo** [comment]
> Question: you also changed `ProtocolClient` by adding schema information. Is this actually needed?

Yes, it is definitely needed actually for all those cases in which you need to be able to convert JSON structure into something different on the wire. This could be a general improvement for several other bindings too.
> Moreover, if we change it we should use `TD.DataSchema` consistently and not mix `TD.DataSchema` and `TD.BaseSchema`.

 I agree, the problem is that for some reasons the compiler complains if I use only DataSchema.

**Comment by danielpeintner** [comment]
Excuse the long idle time due to the holiday season.

>> Moreover, if we change it we should use TD.DataSchema consistently and not mix TD.DataSchema and TD.BaseSchema.
>
> I agree, the problem is that for some reasons the compiler complains if I use only DataSchema.

Mhh, I will try to look into the issue and get back to you.

BTW, the `README.md` states as TODO "tests".
Do you plan to provide tests in this PR? We encourage to have at least some basic tests before merging...

**Comment by lukesmolo** [comment]
I am sorry for this long period of silence, but I am actually busy with other stuff. I added some tests for the binding. In this case, since I cannot instantiate a Netconf server, I made some basic tests.

**Comment by danielpeintner** [comment]
Thanks for the updates.

>>    Question: you also changed ProtocolClient by adding schema information. Is this actually needed?
>
>Yes, it is definitely needed actually for all those cases in which you need to be able to convert JSON structure into something different on the wire. This could be a general improvement for several other bindings too.

 May I ask you again to explain the reasons in more details since we didn't encounter any issues with the current approach. This a _core_ change and we would like to make sure it actually makes sense.

Can you share a concrete example so that we understand the issue better?
Did you look into `ContentSerdes` which is meant to handle different content types.

**Comment by wiresio** [comment]
@lukesmolo I agree with @danielpeintner - would you mind giving us an explanation. Thanks!

**Comment by lukesmolo** [comment]
Sorry again for the delay, I first wanted to discuss with @mkovatsc some details of the bindings. Yes, you are right, for this specific binding it should be feasible to use a new ContentSerdes. In the next days I should be able to work on it and to make a new commit. Thank you for the advice!
PS: Concerning the other OPCUA PR, I will answer on that page since the situation is a little bit different.

**Comment by lukesmolo** [comment]
Ok, now hopefully it should be fine. Maybe the new ContentSerdes in the future could inspire a more generic XML ContentSerdes.

File: packages/binding-netconf/README.md [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+ 
+Please note that, in order to make this binding work, each href should always contain a leaf that refers to a property of the YANG module.
+ 
+## TODO
+ 
+- [ ] Subscriptions implementation (EVENTS)
+- [x] TEST
+- [ ] (NETCONF Server Protocol Binding ?)
+ 
+## Acknowledgments
```
Comment (danielpeintner) [inline packages/binding-netconf/README.md]: Please remove any acknowledgments. We do not have it elsewhere in the project.

File: packages/browser-bundle/package.json [hunk 2]
```diff
@@ -19,8 +19,7 @@
     "browserify": "16.2.3",
     "tinyify": "2.5.0"
   },
-  "dependencies": {
-  },
+  "dependencies": {},
```
Comment (danielpeintner) [inline packages/browser-bundle/package.json]: Please revert this change since this does not belong to this PR.

File: packages/binding-netconf/README.md (hunk 1 above)
Comment (lukesmolo) [reply packages/binding-netconf/README.md]: Ok, done

File: packages/browser-bundle/package.json (hunk 2 above)
Comment (lukesmolo) [reply packages/browser-bundle/package.json]: sure, done

**Comment by danielpeintner** [comment]
@lukesmolo Thank you fort your work!
